tmva/sofie/inc/TMVA/ROperator_Max.hxx

```diff
@@ -0,0 +1,93 @@
+#ifndef TMVA_SOFIE_ROPERATOR_MAX
+#define TMVA_SOFIE_ROPERATOR_MAX
+
+#include "TMVA/SOFIE_common.hxx"
+#include "TMVA/ROperator.hxx"
+#include "TMVA/RModel.hxx"
+
+#include <sstream>
+#include <algorithm>
+
+namespace TMVA{
+namespace Experimental{
+namespace SOFIE{
+
+template <typename T>
+class ROperator_Max final : public ROperator
+{
+
+private:
+
+   std::vector<std::string> fInputNames;
+   std::string fNY;
+   std::vector<std::vector<size_t> > fShapeInputs;
+   std::vector<size_t> fShape;
+
+public:
+   ROperator_Max(){}
+   ROperator_Max( const std::vector<std::string> & inputNames, std::string nameY):
+   fNY(UTILITY::Clean_name(nameY)){
+      fInputNames.reserve(inputNames.size());
+      for (auto & name : inputNames)
+         fInputNames.push_back(UTILITY::Clean_name(name));
+   }
+
+   // type of output given input 
+   std::vector<ETensorType> TypeInference(std::vector<ETensorType> input){
+      return input;
+   }
+
+   // shape of output tensors given input tensors
+   std::vector<std::vector<size_t>> ShapeInference(std::vector<std::vector<size_t>> input){
+      auto ret = std::vector<std::vector<size_t>>(1, input[0]); 
+      return ret;
+   }
+
+   void Initialize(RModel& model){
+      // input must be a graph input, or already initialized intermediate tensor
+
+      for (auto &it : fInputNames) {
+         if (model.CheckIfTensorAlreadyExist(it) == false) {
+            throw std::runtime_error("TMVA SOFIE Concat Op Input Tensor " + it + " is not found in model");
+         }
+         fShapeInputs.push_back(model.GetTensorShape(it));
+      }
+      for(size_t i=0; i < fShapeInputs.size()-1; i++){
+         if(fShapeInputs[i] != fShapeInputs[i+1]){
+            std::string msg = "TMVA SOFIE Max Op: Support only inputs with same shape";
+         throw std::runtime_error(msg);
+         } ;
+      }
+      fShape = fShapeInputs[0];
+      model.AddIntermediateTensor(fNY, model.GetTensorType(fInputNames[0]), fShape);
+      model.AddNeededStdLib("cmath");
+   }
+
+
+   std::string Generate(std::string OpName){
+      OpName = "op_" + OpName;
+      if (fShape.empty()) {
+         throw std::runtime_error("TMVA SOFIE Max called to Generate without being initialized first");
+      }
+      std::stringstream out;
+      size_t length = ConvertShapeToLength(fShape);
+      out << "\n//------ Max\n";
+      out << SP << "for (size_t id = 0; id < " << length << " ; id++){\n";
+      out << SP << SP <<"tensor_" << fNY << "[id] = std::max({";
```

Comment:
It is better here two separate these 3 cases when generating the code: 
- fInputNames.size() == 1 :  `output[id] =  input[0][id]`
- fInputNames.size() == 2 : `output[id] = std::max(input[0][id],input[1][id])`
-  fInputNames.size() > 2 : code as it is now 

By doing this the performances of Max will be optimised for the first two cases.

Reply:
Done with the changes.